python/pyspark/sql/readwriter.py

```diff
@@ -69,20 +83,26 @@ class DataFrameReader(OptionUtils):
     def __init__(self, spark: "SparkSession"):
         self._jreader = spark._jsparkSession.read()
         self._spark = spark
+        self._format: Optional[Union[str, Type[DataSource]]] = None
+        self._schema: Optional[Union[str, StructType]] = None
+        self._options: Dict[str, "OptionalPrimitiveType"] = dict()
 
     def _df(self, jdf: JavaObject) -> "DataFrame":
         from pyspark.sql.dataframe import DataFrame
 
         return DataFrame(jdf, self._spark)
 
-    def format(self, source: str) -> "DataFrameReader":
+    def format(self, source: Union[str, Type[DataSource]]) -> "DataFrameReader":
```

Comment:
can we support a short name?

Reply:
Yes we should also support using short names. But we need to support registering the data source first. This is a TODO in the PR description: `4. Register a Python data source and use its name to reference it in .format()`

Reply:
I actually think also we should remove the support of directly passing `DataSource`. This implementation can't work with short names also if we want to make it working with other languages.

Reply:
Ah you meant we can't have a Python data source object when using it in Scala right? For example this won't work: `val df = spark.read.format(MyPythonDataSource).load()`. This is a valid point. 

@HyukjinKwon just brainstorming here: in order to use the short name, a data source must be registered somewhere. Then we first need to have an API to register a data source, something like `spark.dataSource.register(...)` similar to UDF.  

But can we make it more Pythonic? What if I just want to import a Python data source from a package and directly use it without registering? Having this `DataSource` object we can easily allow this use case:

```python
from src.my.datasources import MyDataSource

df = spark.read.format(MyDataSource).load()
``` 

This is similar to UDFs, where we don't need to register them to use them in PySpark.

Reply:
Yes, we should have a registration mechanism like we do in Scala https://github.com/databricks/spark-xml/blob/master/src/main/resources/META-INF/services/org.apache.spark.sql.sources.DataSourceRegister. 

Can we investigate how other Python libraries do? One way is using INI file like https://packaging.python.org/en/latest/specifications/entry-points/

Reply:
1. Create a INI file either under Spark conf directory, or current working directory (maybe optionally define the location via a static configuration).
2. Read it in `DataSource.lookupDataSource`, and loads the class. We will likely need a wrapper class that inherits Scala Datasource classes.
3. The wrapper Datasource invokes the Python interface to read/write, etc.

Reply:
Discussed offline. We should make it work with short names instead of changing the API here. I will revert the change in this file and work on this part in the next PR. This PR will focus on the execution and API of the data source.